### ISSUE TITLE
Option to output diff

### ISSUE THREAD

**Issue by johansnow** [hashicorp/terraform#23402]
For some resources, such as aws_task_def, having an option to output diff from the previous resource would be a huge gain. We often have terraform create a new task def with very few changes to the container definition, while the output shows create for each of those parameters, which means we have to scroll through roughly 500 lines of output per task - output which is just noise in these cases. If terraform shows destroy on the previous task def, then it also shows destroy for the container definition of the previous task def.

Initially I posted this with aws provider, but they said it relates to Terraform CLI and referred me here. In this case I believe it makes more sense to repost the same information here rather than link to the original issue in their repo.

### Current Terraform Version
```
0.12.13
```

### Use-cases
When Terraform creates a new resource, which mostly contains the same specification as the previous version did.

### Proposal
Turn output such as this (which has been shortened from the actual 500 lines of output)

```
> \# module.base.module.ecs.aws_ecs_task_definition.task_def[9] will be created
> + resource "aws_ecs_task_definition" "task_def" {
>     + arn                      = (known after apply)
>     + container_definitions    = jsonencode(
>           [
>             + {
>                 + cpu               = 240
>                 + dockerLabels      = {
>                     + com.datadoghq.ad.check_names  = jsonencode(
>                           [
>                             + "prometheus",
>                           ]
>                       )
>                     // and a few more - all same as last time, all marked with +
>                   }
>                 + environment       = [
>                     + {
>                         + name  = "service_name"
>                         + value = "example-service"
>                       } 
>                       //  and a lot more environment vars - all same as last time, all marked with +
>                   ]
>                 + secrets           = [
>                     + {
>                         + name      = "foo.bar"
>                         + valueFrom = "arn:aws:ssm:us-east-1:725526741327:parameter/x/foo.bar"
>                       }
>                      // and a lot more more ssm params - all same as last time, all marked with +
>                   ]
>                 + image             = "ACCOUNT-ID.dkr.ecr.us-east-1.amazonaws.com/example-service:IMAGE-TAG"
>                 // Note: "IMAGE-TAG" was the only change between container definition from previous task def to this task def
>               }
>           ]
>       )
> 
```

Into

```
> + resource "aws_ecs_task_definition" "task_def" {
>     + arn                      = (known after apply)
>     + container_definitions    = jsonencode(
>           [
>               - image              = "ACCOUNT-ID.dkr.ecr.us-east-1.amazonaws.com/example-service:PREVIOUS-TAG"
>               + image             = "ACCOUNT-ID.dkr.ecr.us-east-1.amazonaws.com/example-service:IMAGE-TAG"
>           ]
>     )
```

I believe I always want to see only the changes, even when a new resource is created. E.g. in the previous example, there was also output for a new aws_ecs_service resource being created, with the only change being

```
> ~ task_definition                    = "arn:aws:ecs:us-east-1:ACCOUNT-ID:task-definition/x-example-service:258" -> (known after apply)
```

I would **probably** prefer to always see only the changes, but perhaps it would still be a good idea to provide options both for "full output" and "diff output", globally, per module and per resource.

Such an option could be specified after the module/resource name, and on the command line. Deeper nesting overrides previous config, except for default "full output".
E.g.
No output mode specified means default is used, but this default does not override "diff output" config at a higher level.

```
> resource "some_resource" "bar" {
```

Also full output, but since this is explicitly configured, it would override "diff output" at the module level.

```
> resource "some_resource" "baz" full {
```

Diff output

```
> module "foo" diff {
```

So in the above example, if some_resource.bar and some_resource.baz both were defined in module.foo, bar would use diff output and baz would use full output.

Command line
```
> terraform apply --output diff
> terraform apply --output full
```

### References
Initially I opened an issue with AWS provider. It contains some additional background on our usage of aws_task_def, which I chose to strip from this feature request, because even if that particular resource is where this feature would provide us with the biggest benefit, I believe "diff output" would be great to have across the board.
https://github.com/terraform-providers/terraform-provider-aws/issues/10845

**Comment by teamterraform** [comment]
Hi @johansnow! Thanks for sharing this feedback.

We'd like to understand a little better the example situation you're describing here. The initial plan output you shared shows a "create" action, and thus naturally everything inside the block is "new" in the sense that there was no previous object to compare to.

In your second example you show a create action containing an update diff:

```
+ resource "aws_ecs_task_definition" "task_def" {
    + arn                      = (known after apply)
    + container_definitions    = jsonencode(
          [
              - image = "ACCOUNT-ID.dkr.ecr.us-east-1.amazonaws.com/example-service:PREVIOUS-TAG"
              + image = "ACCOUNT-ID.dkr.ecr.us-east-1.amazonaws.com/example-service:IMAGE-TAG"
          ]
    )
```

What we're not sure about is that this diff shows an _old_ value for `image` containing `PREVIOUS-TAG`, but we're not sure where you're intending for that value to come from. Since this is a create action, there is no previous object for that old value to have come from.

In the current implementation, Terraform _can_ show a diff for "replace" actions (either `-/+` or `+/-`) where it will annotate the individual attributes/elements being added, removed, and destroyed, but that is possible only because during a "replace" action there is a previous object to show the difference from.

Would you mind showing the sequence of steps you're taking where you'd expect Terraform to be able to find a previous object to show the diff against? Thanks!

**Comment by mfortin** [comment]
Something similar to this would be awesome and a lot easier to read and process: https://github.com/coinbase/terraform-landscape

**Comment by johansnow** [comment]
Let's make this specific with an example: We are currently on task definition admin-service:1 which has a container definition with about 100 parameters, that results in 400-500 lines of terraform output. One of those parameters defines what docker-image to use:
`"image": "ACCOUNT-ID.dkr.ecr.REGION.amazonaws.com/admin-service:foo"`
We want to change this to instead use 
`"image": "ACCOUNT-ID.dkr.ecr.REGION.amazonaws.com/admin-service:bar"`

We would be perfectly happy if we could introduce this change by updating the container definition of the current task definition. However, that is not how it works. And for good reasons, which is why there are task-definition **revisions**. We are in effect creating a new task definition revision, rather than a completely new and separate task-definition.
Thus, in order to introduce this change, we create a new task definition, admin-service:2. After this is done, a diff of the config JSON of admin-service:1 and admin-service:2 would produce the following output.

```
434c434
<       "image": "ACCOUNT-ID.dkr.ecr.REGION.amazonaws.com/admin-service:foo",
---
>       "image": "ACCOUNT-ID.dkr.ecr.REGION.amazonaws.com/admin-service:bar",
549c549
<   "taskDefinitionArn": "arn:aws:ecs:REGION:ACCOUNT-ID:task-definition/admin-service:1",
---
>   "taskDefinitionArn": "arn:aws:ecs:REGION:ACCOUNT-ID:task-definition/-admin-service:2",
625,626c625,626
<   "revision": 1,
<   "status": "INACTIVE",
---
>   "revision": 2,
>   "status": "ACTIVE",
629,630c629
<   "volumes": [],
<   "statusString": "(INACTIVE)"
---
>   "volumes": []
```

The terraform output however would at the very least show 500 lines of output for the container definition of the new task definition, all with + before them. Potentially, it would also show another 500 lines of output for the container definition of the previous task definition, all with - before them. I'm not entirely certain when the previous task def is shown in the output as destroyed. Perhaps it depends on whether a task definition has been created by some other source than tf or not.

It is certainly true that everything belonging to the old task definition will be removed. I am not saying it is logically false to claim that each of those parameters get destroyed. That is however implied simply by showing a few lines of output for task definition will be destroyed, such as 
`> \# module.base.module.ecs.aws_ecs_task_definition.task_def[9] will be destroyed` followed by one line of output containing `-  PREVIOUS-ARN`

It is certainly true that everything belonging to the new task definition will be created. I am not claiming it is logically false to say that each of those parameters get created.

However, there tends to be a very strong relation between admin-service:1 and admin-service:2 or any admin-service:N and admin-service:N+1 (or N+2 if there has been a non-tf-created task definition in between).

The changes I would like to see in the output
1. When a new task definition is created, I see absolutely no need for being notified about the previous task definition being "destroyed". I created the new one to change a few things in the old one.
2. When a new task definition is created, I see absolutely now need for being notified about everything it contains. For any task definition I create, I am only concerned with the changes introduced as compared to the previous task definition.
3. When the new task definition is the first ever task definition, i.e. `"taskDefinitionArn": "arn:aws:ecs:REGION:ACCOUNT-ID:task-definition/admin-service:1",`
Then the "diff against the previous config would be + for everything, since the previous config didn't exist (was blank, empty).
4. When there is a currently active task definition (admin-service:N), when a new task-definition is created, then it makes perfect sense to only show those parts of the task definition configuration that changed (i.e. diff admin-service:M, the new one, against admin-service:N, the currently active one)

Also, do note: The above suggestions does NOT hold for most resources. For example, if I have one aws_ecs_service and create another, there will be absolutely no connection between those two resources. In this case, it makes perfect sense to produce the full terraform output as is currently done. However, having task definition admin-service:1 and creating admin-service:2 is actually a change to a single aws_ecs_service. And just like it makes sense to show changes to the immediate parameters, such as "desired count" when it changes from 1 to 2, so does it make sense to show changes to the aws_ecs_service's transitive resources, such as the parameters "image" of the container definition belonging to the task definition that belongs to that aws_ecs_service.

**Comment by johansnow** [comment]
I hope my last answer provided enough clarification, but just in case it helps, I actually already have terraform scripts in place that pulls **image** information from aws (meaning the information really is there already, and that there is a connection between the old and new task defs).
> What we're not sure about is that this diff shows an _old_ value for `image` containing `PREVIOUS-TAG`, but we're not sure where you're intending for that value to come from. Since this is a create action, there is no previous object for that old value to have come from.

When it comes specifically to **image**, we are actually changing what docker image is used through a web UI. This is the only container definition parameter we change by other means than terraform. Because of this, we don't want to manage that parameter through terraform. Thus, when we want to update any other parameter, which is done through terraform, we are actually pulling the value for **image** from aws in order to be able to create a new task definition through terraform without risk of introducing this change through terraform

Here is the tf script doing this part
```
data "external" "service_image" {
  count = length(var.services) * (var.enabled ? 1 : 0)

  # output: JSON { "image": "image-url" }
  program = ["/bin/bash", "${path.module}/get-service-image.sh"]
  query = {
    SERVICE = "${var.environment}-${var.services[count.index]["name"]}"
    FIRST_TIME = lookup(
      var.first_time_docker_tags,
      var.services[count.index]["name"],
      "",
      ) == "" ? "" : "${var.services[count.index]["repository"]}:${lookup(
        var.first_time_docker_tags,
        var.services[count.index]["name"],
        "",
    )}"
    TEST     = lookup(var.services[count.index], "test", "")
    NOT      = var.environment == "live" ? "SNAPSHOT$" : ""
    FALLBACK = var.environment == "live" ? "" : "${var.services[count.index]["repository"]}:SNAPSHOT"
  }
}

output "docker-image" {
  value = data.external.service_image.*.result.image
  # value = "${join(",", data.external.service_image.*.result.image)}"
}
```

When it comes to the shell script, disregarding aws session tokens, error handling, and tests, the shell script to pull information from aws regarding the currently active task definition, as well as extracting information about what image is used by that task definition is nothing more than
`image=$(aws ecs describe-task-definition --task-definition $SERVICE | sed -n 's/ *"image":[^"]*"\([^"]*\).*$/\1/gp' | head -n1)`

This value is then inserted into the task-definition that terraform manges.

This is a work-around the fact that we cannot ask terraform to only partially manage our task-definitions. (If possible, we would like it to never ever touch the image parameter of the container definition, but we also don't see this as a big problem, and this is not something I am asking to be changed) 

Our approach works great, with the singular exception that it suffers from the same noisy output as any other terraform created task definition revision.

If we change docker image tag from release-45 to release-46, a new task definition is created. The next time we run terraform, terraform will detect that the task definition has changed because tf state contains a docker image for release-45 (let's say this is task definition admin-service:100), and data is now provided to terraform stating that terraform should create a resource with release-46. In reality (from an aws resource perspective), there is absolutely no need to create a new task definition, because the currently active task definition (task-definition admin-service:101) was created outside of terraform, and will be identical to the task definition terraform now creates (task-definition admin-service:102).

To be perfectly clear, I have no problem with the fact that there is a difference between terraform state and active aws resource, and the fact that task definition revisions 101 and 102 will be identical.

I only dislike the fact that terraform now shows me 500 lines of output with "created status", although the connection between revisions 101 and 102 (as resources belonging to an existing aws_ecs_service)  means it would definitely suffice to say that there was an update + and - for taskDefinitionArn.

Also, to be perfectly clear, the fact that we do manage the docker image outside of terraform is not the only reason I would like to see this behaviour. Even if we stopped doing so, and always made the changes to task def through terraform, the underlying problem of not showing changes between task definitions remain. We would simply get rid of one of the many times we get a lot of noisy output that provide us with absolutely no value from a user-perspective.

**Comment by rvandegrift** [comment]
I'd also like this feature, but not for immutable resources like ECS task definitions (terraform should display them as creates because that's what it is actually doing!  A diff would obscure that.)

For some resources that can be updated in place, `terraform plan` can still be too verbose.  It'd be nice to have an option so I only see the changed properties.

Here is a no-change plan that results from an upgrade to the GCP provider.  The length makes it hard to find the significant lines.
```
  # google_container_cluster.nonprod will be updated in-place
  ~ resource "google_container_cluster" "nonprod" {
      ~ additional_zones            = [
          - "us-central1-a",
          - "us-central1-b",
          - "us-central1-c",
        ]
        cluster_ipv4_cidr           = "10.12.0.0/14"
        default_max_pods_per_node   = 110
        enable_binary_authorization = false
        enable_intranode_visibility = false
        enable_kubernetes_alpha     = false
        enable_legacy_abac          = false
        enable_shielded_nodes       = true
        enable_tpu                  = false
        endpoint                    = "192.0.2.1"
        id                          = "nonprod"
        initial_node_count          = 0
        instance_group_urls         = [
            "https://www.googleapis.com/compute/beta/projects/fake-project/zones/us-central1-a/instanceGroups/gke-nonprod-core-nodes-2019111401-320adb6c-grp",
            "https://www.googleapis.com/compute/beta/projects/fake-project/zones/us-central1-b/instanceGroups/gke-nonprod-core-nodes-2019111401-16e98f97-grp",
            "https://www.googleapis.com/compute/beta/projects/fake-project/zones/us-central1-c/instanceGroups/gke-nonprod-core-nodes-2019111401-28c5c7bb-grp",
        ]
        location                    = "us-central1"
        logging_service             = "logging.googleapis.com/kubernetes"
        master_version              = "1.13.11-gke.15"
        min_master_version          = "1.13.11-gke.14"
        monitoring_service          = "monitoring.googleapis.com/kubernetes"
        name                        = "nonprod"
        network                     = "projects/fake-project/global/networks/fake-project"
        node_locations              = [
            "us-central1-a",
            "us-central1-b",
            "us-central1-c",
        ]
        node_version                = "1.13.11-gke.14"
        project                     = "fake-project"
      - region                      = "us-central1" -> null
        remove_default_node_pool    = true
        resource_labels             = {}
        services_ipv4_cidr          = "10.16.192.0/20"
        subnetwork                  = "projects/fake-project/regions/us-central1/subnetworks/npc-guc1"

        addons_config {

            istio_config {
                auth     = "AUTH_MUTUAL_TLS"
                disabled = false
            }

            network_policy_config {
                disabled = false
            }
        }

        cluster_autoscaling {
            enabled = false
        }

        database_encryption {
            key_name = "projects/fake-project/locations/us-central1/keyRings/gke_nonprod/cryptoKeys/etcd_db_key"
            state    = "ENCRYPTED"
        }

        ip_allocation_policy {
            cluster_ipv4_cidr_block       = "10.12.0.0/14"
            cluster_secondary_range_name  = "k8s-cluster"
            create_subnetwork             = false
            services_ipv4_cidr_block      = "10.16.192.0/20"
            services_secondary_range_name = "k8s-services"
            use_ip_aliases                = false
        }

        maintenance_policy {
            daily_maintenance_window {
                duration   = "PT4H0M0S"
                start_time = "04:00"
            }
        }

        master_auth {
            cluster_ca_certificate = "REDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTEDREDACTED"

            client_certificate_config {
                issue_client_certificate = false
            }
        }

        master_authorized_networks_config {
            cidr_blocks {
                cidr_block   = "192.0.2.16/28"
                display_name = "permit-example"
            }
        }

        network_policy {
            enabled  = true
            provider = "CALICO"
        }

        node_config {
            disk_size_gb      = 100
            disk_type         = "pd-standard"
            guest_accelerator = []
            image_type        = "COS"
            labels            = {}
            local_ssd_count   = 0
            machine_type      = "n1-standard-4"
            metadata          = {
                "disable-legacy-endpoints" = "true"
            }
            oauth_scopes      = [
                "https://www.googleapis.com/auth/logging.write",
                "https://www.googleapis.com/auth/monitoring",
            ]
            preemptible       = false
            service_account   = "nonprod-cluster@fake-project.iam.gserviceaccount.com"
            tags              = []
            taint             = []

            shielded_instance_config {
                enable_integrity_monitoring = true
                enable_secure_boot          = true
            }

            workload_metadata_config {
                node_metadata = "GKE_METADATA_SERVER"
            }
        }

        node_pool {
            initial_node_count  = 1
            instance_group_urls = [
                "https://www.googleapis.com/compute/v1/projects/fake-project/zones/us-central1-a/instanceGroupManagers/gke-nonprod-core-nodes-2019111401-320adb6c-grp",
                "https://www.googleapis.com/compute/v1/projects/fake-project/zones/us-central1-b/instanceGroupManagers/gke-nonprod-core-nodes-2019111401-16e98f97-grp",
                "https://www.googleapis.com/compute/v1/projects/fake-project/zones/us-central1-c/instanceGroupManagers/gke-nonprod-core-nodes-2019111401-28c5c7bb-grp",
            ]
            max_pods_per_node   = 110
            name                = "core-nodes-2019111401"
            node_count          = 1
            node_locations      = [
                "us-central1-a",
                "us-central1-b",
                "us-central1-c",
            ]
            version             = "1.13.11-gke.14"

            autoscaling {
                max_node_count = 3
                min_node_count = 1
            }

            management {
                auto_repair  = true
                auto_upgrade = true
            }

            node_config {
                disk_size_gb      = 100
                disk_type         = "pd-standard"
                guest_accelerator = []
                image_type        = "COS"
                labels            = {}
                local_ssd_count   = 0
                machine_type      = "n1-standard-4"
                metadata          = {
                    "disable-legacy-endpoints" = "true"
                }
                oauth_scopes      = [
                    "https://www.googleapis.com/auth/logging.write",
                    "https://www.googleapis.com/auth/monitoring",
                ]
                preemptible       = false
                service_account   = "nonprod-cluster@fake-project.iam.gserviceaccount.com"
                tags              = []
                taint             = []

                shielded_instance_config {
                    enable_integrity_monitoring = true
                    enable_secure_boot          = true
                }

                workload_metadata_config {
                    node_metadata = "GKE_METADATA_SERVER"
                }
            }
        }

        private_cluster_config {
            enable_private_endpoint = false
            enable_private_nodes    = true
            master_ipv4_cidr_block  = "172.27.208.64/28"
            peering_name            = "gke-d30a55f8149cf14b09d4-8f3b-c258-peer"
            private_endpoint        = "172.27.208.66"
            public_endpoint         = "192.0.2.1"
        }

        release_channel {
            channel = "UNSPECIFIED"
        }

        workload_identity_config {
            identity_namespace = "fake-project.svc.id.goog"
        }
    }

```

The proposed `terraform plan -output=diff` view makes it obvious what changes:
```
  # google_container_cluster.nonprod will be updated in-place
  ~ resource "google_container_cluster" "nonprod" {
      ~ additional_zones            = [
          - "us-central1-a",
          - "us-central1-b",
          - "us-central1-c",
      - region                      = "us-central1" -> null
    }
```

You can get close with grep, but it's hard to get the nice format with closing braces and status messages.  (After a quick think, the best I could do: `terraform plan | grep -E '^([[:space:]]*[#~+-]|[[:alnum:]])'`)

The full output should still be the default.

**Comment by alisdair** [comment]
Thanks to everyone who contributed to this discussion, especially for the many clear use cases for a shorter diff!

Our approach to addressing this problem was recently merged in #26187, and shipped today in an alpha prerelease. If you're able to do so, I would encourage you to try out the new diff renderer in the alpha build, and [leave feedback on our discussion forum thread for this topic](https://discuss.hashicorp.com/t/terraform-0-14-concise-plan-diff-feedback-thread/14004).

In brief, the approach we hope to release as part of 0.14.0:

- Renders only changes and identifying attributes, hiding unchanged attributes and blocks, which makes diffs much shorter
- Keeps enough context to ensure that the diff can be understood in isolation
- Does not require a CLI switch or configuration setting, to make it more easily discoverable

We hope that this changed approach to rendering plans will solve the core problems discussed here, so I'm going to close this issue for now. If you have feedback on the new approach, please [post in the discussion thread](https://discuss.hashicorp.com/t/terraform-0-14-concise-plan-diff-feedback-thread/14004). Looking forward to reading your thoughts!
